external-crates/move/move-compiler/src/typing/translate.rs

```diff
@@ -1624,13 +1611,28 @@ fn loop_body(
         eloop.ty.clone(),
         sp(lloc, Type_::Unit),
     );
-    let has_break = break_type_opt.is_some();
-    let ty = if is_loop && !has_break {
-        core::make_tvar(context, lloc)
+
+    let break_ty_opt = context.named_block_type_opt(name);
+
+    if let Some(break_ty) = break_ty_opt {
+        if !is_loop {
+            // while loop breaks must also be unit
+            subtype(
+                context,
+                break_ty.loc,
+                || "Cannot use 'break' with a value 'while'",
```

Comment:
Wording is a bit off maybe, because you can... as long as it is `()` :p 

While it isn't really a value, it kind of is from the source languages perspective